### ISSUE TITLE
data parallelism in keras with Tensorflow

### ISSUE THREAD

**Issue by volvador** [fchollet/keras#5405]
I am confused by the explanation of distributed training here :
https://blog.keras.io/keras-as-a-simplified-interface-to-tensorflow-tutorial.html

Let us say I have this simple model
x_train, y_train = .... # my daya

model = Sequential()
model.add(Dense(output_dim = 10,  input_dim = x_train.shape[1]))
model.add(Activation('tanh'))
model.add(Dropout(0.5))
model.add(Dense(output_dim = y_train.shape[1], input_dim = 10))
model.compile(loss='mse', optimizer='rmsprop')

I have a machine with 8 CPU (not GPU) cores and I want to distribute my training on the 8 cores. I want data parallelism and not model parallelism.

According to the explanation in the keras blog, I have only to do this:
server = tf.train.Server.create_local_server()
sess = tf.Session(server.target)
from keras import backend as K
K.set_session(sess)

and then call
model.fit(X_train, y_train, nb_epoch=20, batch_size=200)

But I do not see how tf will understand that I want to distribute my computations on the 8 CPU cores, doing minibatches of 25 in each of them and averaging the SGD step.

Also, @fchollet  gives a more elaborate manner to do data parallelism here
https://www.quora.com/What-is-the-state-of-distributed-learning-multi-GPU-and-across-multiple-hosts-in-Keras-and-what-are-the-future-plans

which seems much more complicated than what is say in the keras blog, adding to my confusion.

TL;DR for the model defined above, and having 8 CPU cores on my machine, how can I do distributed training (data parallelism) using keras/Tensorflow?

**Comment by laxatives** [comment]
I'm not sure I understand the details of your question, but `tf.Session(config=tf.ConfigProto(intra_op_parallelism_threads=8))` should be sufficient to have your session execute training on 8 threads on a single machine.

**Comment by volvador** [comment]
@laxatives Thanks for your answer.
My understanding is that the config option you suggest enables Eigen operations to be done on multiple threads. But the learning is still sequential. There is no gradient estimation done in parallel on different batches of data (each core replicating the model, and computing the gradient on a different minibatch of data). It is precisely this parallelism that I need. Any idea on that?

**Comment by stone8oy** [comment]
By the way,how to run keras data parallel on multiple machines ? Any idea on this ? Thanks!

**Comment by vagarwal87** [comment]
did anyone find out the answer to this? I had the same question

**Comment by PeterPanUnderhill** [comment]
I have the same questions here. Since I read from some dicussions that Keras automatically uses all cores available, whereis here it seems Tensorflow uses only one core. So what is the combined effect? All cores used?

**Comment by hainguyenct** [comment]
I have the same problem. I declared 10 cpu cores for the job, but only a few (2 or 3) of them worked with 100%

**Comment by siddheshsathe** [comment]
You can have a backend from tensorflow here.
```
import keras.backend.tensorflow_backend as K
with K.tf.device('/<your device>'): # Ex. can be /gpu:0 or /cpu:0 or /device:GPU:2 for multigpu system
        K.set_session(K.tf.Session(config=K.tf.ConfigProto(allow_soft_placement=True, log_device_placement=True)))
        model = Sequential()
        # Your model
        # Compile model
        model.summary()
        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
```

**Comment by shakthydoss** [comment]
> You can have a backend from tensorflow here.
> 
> ```
> import keras.backend.tensorflow_backend as K
> with K.tf.device('/<your device>'): # Ex. can be /gpu:0 or /cpu:0 or /device:GPU:2 for multigpu system
>         K.set_session(K.tf.Session(config=K.tf.ConfigProto(allow_soft_placement=True, log_device_placement=True)))
>         model = Sequential()
>         # Your model
>         # Compile model
>         model.summary()
>         model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
> ```
What does cpu:0 mean ? 
How many CPUs it will use.

**Comment by siddheshsathe** [comment]
> > You can have a backend from tensorflow here.
> > ```
> > import keras.backend.tensorflow_backend as K
> > with K.tf.device('/<your device>'): # Ex. can be /gpu:0 or /cpu:0 or /device:GPU:2 for multigpu system
> >         K.set_session(K.tf.Session(config=K.tf.ConfigProto(allow_soft_placement=True, log_device_placement=True)))
> >         model = Sequential()
> >         # Your model
> >         # Compile model
> >         model.summary()
> >         model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
> > ```
> 
> What does cpu:0 mean ?
> How many CPUs it will use.

cpu:0 means the very first CPU. If your device as multiple,  you can put the jobs on them as well.
<br>
Similar applies to the GPU. If your machine has several GPUs connected, those will be enumerated from gpu:0 through gpu:n.
<br>
For number of cpus in your machine, run below command to see:
# cpuinfo
